packages/shared/pkg/storage/storage_google.go

```diff
@@ -210,23 +217,61 @@ func (g *GCPBucketStorageObjectProvider) WriteTo(dst io.Writer) (int64, error) {
 }
 
 func (g *GCPBucketStorageObjectProvider) WriteFromFileSystem(path string) error {
-	cmd := exec.CommandContext(
-		g.ctx,
-		"gcloud",
-		"storage",
-		"cp",
-		"--verbosity",
-		"error",
-		path,
-		fmt.Sprintf("gs://%s/%s", g.storage.bucket.BucketName(), g.path),
-	)
+	upload := func() error {
+		systemRunArgs := []string{
+			"--user",
+			"--scope",
+		}
+		if g.limiter != nil {
+			uploadLimiter := g.limiter.GCloudUploadLimiter()
+			if uploadLimiter != nil {
+				semaphoreErr := uploadLimiter.Acquire(g.ctx, 1)
+				if semaphoreErr != nil {
+					return fmt.Errorf("failed to acquire semaphore: %w", semaphoreErr)
+				}
+				defer uploadLimiter.Release(1)
+			}
+
+			systemRunArgs = append(systemRunArgs, g.limiter.GCloudCmdLimits(path)...)
+		}
 
-	output, err := cmd.CombinedOutput()
-	if err != nil {
-		return fmt.Errorf("failed to upload file to GCS: %w\n%s", err, string(output))
+		args := append(systemRunArgs,
+			"gcloud",
+			"storage",
+			"cp",
+			"--verbosity", "error",
+			path,
+			fmt.Sprintf("gs://%s/%s", g.storage.bucket.BucketName(), g.path),
+		)
+
+		cmd := exec.CommandContext(
+			g.ctx,
+			"systemd-run",
```

Comment:
I'm not sure, but we might want to add `--scope` here if `cmd.CombinedOutput()` no longer contains stdout/stderr. It's not super clear to me, but if we're still receiving output, feel free to ignore this!